### ISSUE TITLE
bug(profile): firestore profile causes error: "parsing string to date. JSON parse error: Unexpected identifier undefined"

### ISSUE THREAD

**Issue by compojoom** [prescottprue/react-redux-firebase#391]
**Do you want to request a *feature* or report a *bug*?**  
bug

**What is the current behavior?**
react-redux-firebase 2.0.3
redux-firestore - 0.2.6
react-native-firebase 3.2.3

As reported on gitter by @krewllobster creating a user account and then populating his profile causes an error: error parsing string to date. JSON parse error: Unexpected identifier "undefined"

when the profile should be populated on firestore. I had a look at the code in here:
https://github.com/prescottprue/react-redux-firebase/blob/master/src/actions/auth.js#L207

But it doesn't make sense to me.

Above that line we do:
```
  // use profileFactory if it exists in config
  if (isFunction(config.profileFactory)) {
    // catch errors in user provided profileFactory function
    try {
      profile = config.profileFactory(userData, profile) // eslint-disable-line no-param-reassign
    } catch (err) {
      /* eslint-disable no-console */
      console.error(
        'Error occured within profileFactory function:',
        err.message || err
      )
      /* eslint-enable no-console */
      return Promise.reject(err)
    }
  }
```

If the user has specified a profileFactory - why are we ignoring it? and going with the userData returned from auth? Why aren't we using what the profileFactory is generating? If I specify a profile factory I would expect that you would use it 1:1 and wouldn't alter the result?

Why are we adding craetedAT, lastLogingAt, avatarUrl - if those are available on the auth object?

Leaving that aside the problem seems to come from the fact that SET_PROFILE is called twice, but the first time it's null
<img width="901" alt="grafik" src="https://user-images.githubusercontent.com/693770/35672602-6218d23c-073f-11e8-9acd-db2b4b446409.png">

And it stringToDate can't cope with that.
@prescottprue - what do you suggest we should do here? For me the current behavior seems to be wrong. I see you've added the createdAt etc in 2.0.2, but I don't understand why :)

cheers!

**Comment by krewllobster** [comment]
To add one more piece, when createUser() is called, it first calls the login() method. This seems to ignore my reduxFirebaseConfig that specifies resetBeforeLogin = false, and dispatches the LOGIN_ERROR action regardless. This occurs here: https://github.com/prescottprue/react-redux-firebase/blob/master/src/actions/auth.js#L456

@compojoom thanks for writing this up!

**Comment by prescottprue** [comment]
Unless I am misunderstanding your explanation, `SET_PROFILE` is supposed to be called twice (once to clear state when starting auth and once to set it after profile result comes back).

As for this
> If the user has specified a profileFactory - why are we ignoring it? and going with the userData returned from auth?
...
> I see you've added the createdAt etc in 2.0.2, but I don't understand why :)

By default `userData` includes `createdAt` which is a timestamp, which was causing issues when not providing a `profileFactory` that handled this. In switching to handling the date logic internally it looks like we dropped support for `profileFactory` with Firestore all together (not intended).

**Comment by compojoom** [comment]
Ok, i'll check again if the null on set_profile is causing the issue.
In my case userData didn't container createdAt - createdat was available in the metadata property.

I personally don't like the internal handling of this. I think that it should be disabled completely if you specify profile factory. I'm not storing created_at and last login on the user profile. And also there are a lot of unnecessary properties saved on it right now - photoUrl etc.

**Comment by prescottprue** [comment]
@compojoom I agree with not liking the internal handling. It is nice to have a default that works for those that want to try things without specifying anything, but it should be completely configurable (including not running at all if you provide your own factory).

As for how to handle it, we should move [this logic handling converting user data](https://github.com/prescottprue/react-redux-firebase/blob/master/src/actions/auth.js#L215-L223) to its own function that only runs if `profileFactory` does not exist.

**Comment by prescottprue** [comment]
Included in [`v2.0.4` release](https://github.com/prescottprue/react-redux-firebase/releases/tag/v2.0.4). Thanks for reporting and providing the PR!
